app/models/spree/order_decorator.rb

```diff
@@ -0,0 +1,26 @@
+module Spree
+  module OrderDecorator
+    def self.prepended(base)
+      return if base.checkout_steps.key?(:payment_confirm)
+
+      base.insert_checkout_step(
+        :payment_confirm,
+        before: :complete,
+        if: lambda do |order|
+          order.payments.valid.map { |p| p.payment_method&.has_preference?(:intents) && p.payment_method&.get_preference(:intents) }.any?
+        end
+      )
+    end
+
+    def process_payments!
+      # Payments are processed in confirm_payment step where after successful
+      # 3D Secure authentication `intent_client_key` is saved for payment.
+      # In case authentication is unsuccessful, `intent_client_key` is removed.
+      return unless payments.last.intent_client_key.nil?
```

Comment:
```suggestion
      return if payments.valid.blank?
      return if payments.valid.last.intent_client_key.blank?
```

there will be situations in which payment isn't needed (eg. free orders)

Reply:
In the case of free order `order.payment_required?` is `false` so it won't go to the `payment` state (https://github.com/spree/spree/blob/d2a2dc66eda5798df93d8ca4f8f310c13dc69678/core/app/models/spree/order.rb#L52).
If there is no `payment` state, the order payments (which are not there) won't be processed before the `complete` state (https://github.com/spree/spree/blob/d2a2dc66eda5798df93d8ca4f8f310c13dc69678/core/app/models/spree/order/checkout.rb#L77).
What am I missing?
Are there any other cases that should be included here?